File: modules/stackr-card-vision/android/src/main/java/com/stackr/cardvision/StackrCardFrameAnalyserInstrumentation.kt
package com.stackr.cardvision

import android.os.Handler
import android.os.Looper
import java.lang.ref.WeakReference
import java.util.concurrent.atomic.AtomicLong
import kotlin.math.ceil
import kotlin.math.max

internal const val STACKR_CARD_FRAME_ANALYSIS_EVENT = "onCardFrameAnalysis"

internal object StackrCardFrameAnalyserInstrumentation {
  private const val maxSamples = 80
  private val mainHandler = Handler(Looper.getMainLooper())
  private var moduleReference: WeakReference<StackrCardVisionModule>? = null
  private val received = AtomicLong(0)
  private val processed = AtomicLong(0)
  private val dropped = AtomicLong(0)
  private val focusFailures = AtomicLong(0)
  private val durations = ArrayDeque<Double>()

  @Synchronized
  fun attach(module: StackrCardVisionModule) {
    moduleReference = WeakReference(module)
  }

  @Synchronized
  fun detach(module: StackrCardVisionModule) {
    if (moduleReference?.get() === module) {
      moduleReference = null
    }
  }

  @Synchronized
  fun reset() {
    received.set(0)
    processed.set(0)
    dropped.set(0)
    focusFailures.set(0)
    durations.clear()
  }

  fun markReceived(): Long = received.incrementAndGet()

  fun markDropped(): Long = dropped.incrementAndGet()

  fun markFocusFailure(): Long = focusFailures.incrementAndGet()

  @Synchronized
  fun emitResult(scanId: String?, result: CardFrameAnalysisResult) {
    processed.incrementAndGet()
    durations.addLast(result.processingMs)
    while (durations.size > maxSamples) {
      durations.removeFirst()
    }
    val payload = mapOf(
      "scanId" to scanId,
      "analysisFramesReceived" to received.get(),
      "framesProcessed" to processed.get(),
      "framesDropped" to dropped.get(),
      "focusFailures" to focusFailures.get(),
      "analyserP50Ms" to percentile(durations.toList(), 0.5),
      "analyserP95Ms" to percentile(durations.toList(), 0.95),
      "result" to result.toMap()
    )
    emit(payload)
  }

  fun emitAvailability(scanId: String?, message: String) {
    emit(
      mapOf(
        "scanId" to scanId,
        "analysisFramesReceived" to received.get(),
        "framesProcessed" to processed.get(),
        "framesDropped" to dropped.get(),
        "focusFailures" to focusFailures.get(),
        "analyserP50Ms" to percentileSafe(),
        "analyserP95Ms" to percentileSafe(0.95),
        "result" to null,
        "message" to message
      )
    )
  }

  @Synchronized
  fun snapshot(): Map<String, Any?> = mapOf(
    "analysisFramesReceived" to received.get(),
    "framesProcessed" to processed.get(),
    "framesDropped" to dropped.get(),
    "focusFailures" to focusFailures.get(),
    "analyserP50Ms" to percentile(durations.toList(), 0.5),
    "analyserP95Ms" to percentile(durations.toList(), 0.95)
  )

  private fun emit(payload: Map<String, Any?>) {
    val module = moduleReference?.get() ?: return
    mainHandler.post {
      module.sendEvent(STACKR_CARD_FRAME_ANALYSIS_EVENT, payload)
    }
  }

  @Synchronized
  private fun percentileSafe(ratio: Double = 0.5): Double = percentile(durations.toList(), ratio)

  private fun percentile(values: List<Double>, ratio: Double): Double {
    if (values.isEmpty()) return 0.0
    val sorted = values.sorted()
    val index = (ceil(sorted.size * ratio).toInt() - 1).coerceIn(0, max(0, sorted.size - 1))
    return sorted[index]
  }
}

File: modules/stackr-card-vision/ios/StackrCardFrameAnalyserInstrumentation.swift
import Foundation

let stackrCardFrameAnalysisEvent = "onCardFrameAnalysis"

enum StackrCardFrameAnalyserInstrumentation {
  private static let maxSamples = 80
  private static let lock = NSLock()
  private static weak var module: StackrCardVisionModule?
  private static var received: Int64 = 0
  private static var processed: Int64 = 0
  private static var dropped: Int64 = 0
  private static var focusFailures: Int64 = 0
  private static var durations: [Double] = []

  static func attach(_ module: StackrCardVisionModule) {
    lock.lock()
    self.module = module
    lock.unlock()
  }

  static func detach(_ module: StackrCardVisionModule) {
    lock.lock()
    if self.module === module {
      self.module = nil
    }
    lock.unlock()
  }

  static func reset() {
    lock.lock()
    received = 0
    processed = 0
    dropped = 0
    focusFailures = 0
    durations.removeAll()
    lock.unlock()
  }

  @discardableResult
  static func markReceived() -> Int64 {
    lock.lock()
    received += 1
    let value = received
    lock.unlock()
    return value
  }

  @discardableResult
  static func markDropped() -> Int64 {
    lock.lock()
    dropped += 1
    let value = dropped
    lock.unlock()
    return value
  }

  @discardableResult
  static func markFocusFailure() -> Int64 {
    lock.lock()
    focusFailures += 1
    let value = focusFailures
    lock.unlock()
    return value
  }

  static func emitResult(scanId: String?, result: CardFrameAnalysisResult) {
    lock.lock()
    processed += 1
    durations.append(result.processingMs)
    if durations.count > maxSamples {
      durations.removeFirst(durations.count - maxSamples)
    }
    let payload = snapshotLocked(scanId: scanId, result: result.toDictionary(), message: nil)
    let targetModule = module
    lock.unlock()

    DispatchQueue.main.async {
      targetModule?.sendEvent(stackrCardFrameAnalysisEvent, payload)
    }
  }

  static func emitAvailability(scanId: String?, message: String) {
    lock.lock()
    let payload = snapshotLocked(scanId: scanId, result: nil, message: message)
    let targetModule = module
    lock.unlock()

    DispatchQueue.main.async {
      targetModule?.sendEvent(stackrCardFrameAnalysisEvent, payload)
    }
  }

  static func snapshot() -> [String: Any?] {
    lock.lock()
    let payload = snapshotLocked(scanId: nil, result: nil, message: nil)
    lock.unlock()
    return payload
  }

  private static func snapshotLocked(
    scanId: String?,
    result: [String: Any?]?,
    message: String?
  ) -> [String: Any?] {
    var payload: [String: Any?] = [
      "scanId": scanId,
      "analysisFramesReceived": received,
      "framesProcessed": processed,
      "framesDropped": dropped,
      "focusFailures": focusFailures,
      "analyserP50Ms": percentile(durations, ratio: 0.5),
      "analyserP95Ms": percentile(durations, ratio: 0.95)
    ]
    if result != nil {
      payload["result"] = result
    }
    if message != nil {
      payload["message"] = message
    }
    return payload
  }

  private static func percentile(_ values: [Double], ratio: Double) -> Double {
    if values.isEmpty { return 0 }
    let sorted = values.sorted()
    let index = max(0, min(sorted.count - 1, Int(ceil(Double(sorted.count) * ratio)) - 1))
    return sorted[index]
  }
}
